File: LeetCode/search_a_2d_matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if(!matrix.size() || !matrix[0].size())
            return false;
        
        int i = 0, j = matrix.size()-1;
        int mid = 0;
        
        while(i <= j) {
            mid = (i + j) / 2;
            
            if(matrix[mid][0] == target)
                return true;
            
            if(matrix[mid][0] > target)
                j = mid-1;
            else
                i = mid+1;
        }
        
        int line;
        
        if(mid > 0)
            line = matrix[mid][0] > target ? mid-1 : mid;
        
        i = 0; j = matrix[line].size()-1;
        
        while(i <= j) {
            mid = (i + j) / 2;
            
            if(matrix[line][mid] == target)
                return true;
            
            if(matrix[line][mid] > target)
                j = mid-1;
            else
                i = mid+1;
        }
        
        return false;
    }
};

File: LeetCode/search_a_2d_matrix.go
func searchMatrix(matrix [][]int, target int) bool {
    l, r := 0, len(matrix)-1

    rowPos := 0
    for l <= r {
        mid := (l + r) / 2

        if matrix[mid][0] == target {
            return true
        }

        if matrix[mid][0] < target {
            rowPos = mid
            l = mid + 1
        } else {
            r = mid - 1
        }
    }

    l, r = 0, len(matrix[rowPos])-1
    for l <= r {
        mid := (l + r) / 2

        if matrix[rowPos][mid] == target {
            return true
        }

        if matrix[rowPos][mid] < target {
            l = mid + 1
        } else {
            r = mid - 1
        }
    }

    return false
}
